src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -95,28 +85,121 @@ public class PublishPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
+
+                if(!PipelineArtifactPathHelper.IsValidArtifactName(artifactName)) {
+                    throw new ArgumentException(StringUtil.Loc("ArtifactNameIsNotValid", artifactName));
+                }
+
+                string fullPath = Path.GetFullPath(targetPath);
+                bool isFile = File.Exists(fullPath);
+                bool isDir = Directory.Exists(fullPath);
+                if (!isFile && !isDir)
+                {
+                    // if local path is neither file nor folder
+                    throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
+                }
+
+                // Upload to VSTS BlobStore, and associate the artifact with the build.
+                context.Output(StringUtil.Loc("UploadingPipelineArtifact", fullPath, buildId));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.UploadAsync(context, projectId, buildId, artifactName, fullPath, token);
+                context.Output(StringUtil.Loc("UploadArtifactFinished"));
 
-            string fullPath = Path.GetFullPath(targetPath);
-            bool isFile = File.Exists(fullPath);
-            bool isDir = Directory.Exists(fullPath);
-            if (!isFile && !isDir)
-            {
-                // if local path is neither file nor folder
-                throw new FileNotFoundException(StringUtil.Loc("PathDoesNotExist", targetPath));
             }
+            else if (artifactType == fileShareType)
+            {
+                string fileSharePath = context.GetInput(ArtifactEventProperties.FileSharePath, required: true);
+                string artifactPath = Path.Join(fileSharePath, artifactName);
+
+                if(RuntimeInformation.IsOSPlatform(OSPlatform.Windows)){
+                    // create the artifact. at this point, mkdirP already succeeded so the path is good.
+                    // the artifact should get cleaned up during retention even if the copy fails in the
+                    // middle
+                    Directory.CreateDirectory(artifactPath);
+
+                    // 2) associate the pipeline artifact with an build artifact
+                    VssConnection connection = context.VssConnection;
+                    BuildServer buildHelper = new BuildServer(connection);
+                    Dictionary<string, string> propertiesDictionary = new Dictionary<string, string>();
+                    propertiesDictionary.Add(FileShareArtifactUploadEventProperties.ArtifactName, artifactName);
+                    propertiesDictionary.Add(FileShareArtifactUploadEventProperties.ArtifactType, fileShareType);
+                    propertiesDictionary.Add(FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath);
+
+                    var artifact = await buildHelper.AssociateArtifact(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
```

Comment:
>AssociateArtifact [](start = 53, length = 17)

I know you didn't change this but all the BuildServer methods names need to end with "Async" since they are async methods.

Reply:
Yeah, doesn't hurt to improve the code base though.